File: internal/domain/stats.go
package domain

import "time"

// StatsSummary holds aggregated statistics for a time period.
type StatsSummary struct {
	TotalQueries  int64   `json:"totalQueries"`
	BlockedCount  int64   `json:"blockedCount"`
	CachedCount   int64   `json:"cachedCount"`
	BlockedPercent float64 `json:"blockedPercent"`
	CacheHitRatio float64 `json:"cacheHitRatio"`
}

// StatsPoint represents a single data point in a time series.
type StatsPoint struct {
	Timestamp    time.Time `json:"timestamp"`
	TotalQueries int64     `json:"totalQueries"`
	BlockedCount int64     `json:"blockedCount"`
	CachedCount  int64     `json:"cachedCount"`
}

// TopDomain represents a domain ranked by query count.
type TopDomain struct {
	Domain string `json:"domain"`
	Count  int64  `json:"count"`
}

// TopClient represents a client ranked by query volume.
type TopClient struct {
	IP       string `json:"ip"`
	Hostname string `json:"hostname,omitempty"`
	Count    int64  `json:"count"`
}

// StatBucket holds aggregated stats for a time bucket (hour/day).
type StatBucket struct {
	Timestamp    time.Time          `json:"timestamp"`
	TotalQueries int64              `json:"totalQueries"`
	BlockedCount int64              `json:"blockedCount"`
	CachedCount  int64              `json:"cachedCount"`
	TopDomains   map[string]int64   `json:"topDomains,omitempty"`
	TopClients   map[string]int64   `json:"topClients,omitempty"`
}

File: web/src/api/stats.ts
import { useQuery } from '@tanstack/react-query'
import { apiClient } from './client'
import { REFETCH_INTERVAL } from '../utils/constants'

interface StatsSummary {
  totalQueries: number
  blockedCount: number
  cachedCount: number
  blockedPercent: number
  cacheHitRatio: number
}

interface StatsPoint {
  timestamp: string
  totalQueries: number
  blockedCount: number
  cachedCount: number
}

interface TopDomain {
  domain: string
  count: number
}

interface TopClient {
  ip: string
  hostname?: string
  count: number
}

export function useStatsSummary() {
  return useQuery({
    queryKey: ['stats', 'summary'],
    queryFn: () => apiClient.get<{ data: StatsSummary }>('/stats/summary').then((r) => r.data),
    refetchInterval: REFETCH_INTERVAL,
  })
}

export function useStatsOvertime() {
  return useQuery({
    queryKey: ['stats', 'overtime'],
    queryFn: () => apiClient.get<{ data: StatsPoint[] }>('/stats/overtime').then((r) => r.data),
    refetchInterval: REFETCH_INTERVAL,
  })
}

export function useTopDomains() {
  return useQuery({
    queryKey: ['stats', 'top-domains'],
    queryFn: () => apiClient.get<{ data: { allowed: TopDomain[]; blocked: TopDomain[] } }>('/stats/top-domains').then((r) => r.data),
    refetchInterval: REFETCH_INTERVAL,
  })
}

export function useTopClients() {
  return useQuery({
    queryKey: ['stats', 'top-clients'],
    queryFn: () => apiClient.get<{ data: TopClient[] }>('/stats/top-clients').then((r) => r.data),
    refetchInterval: REFETCH_INTERVAL,
  })
}
